python/tests/ibd.py

```diff
@@ -0,0 +1,421 @@
+# MIT License
+#
+# Copyright (c) 2020 Tskit Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Python implementation of the IBD-finding algorithms.
+"""
+import argparse
+
+import tskit
+
+
+class Segment:
+    """
+    A class representing a single segment. Each segment has a left and right,
+    denoting the loci over which it spans, a node and a next, giving the next
+    in the chain.
+
+    The node it records is the *output* node ID.
+    """
+
+    def __init__(self, left=None, right=None, node=None, next_seg=None):
+        self.left = left
+        self.right = right
+        self.node = node
+        self.next = next_seg
+
+    def __str__(self):
+        s = "({}-{}->{}:next={})".format(
+            self.left, self.right, self.node, repr(self.next)
+        )
+        return s
+
+    def __repr__(self):
+        return repr((self.left, self.right, self.node))
+
+    def __eq__(self, other):
+        # NOTE: to simplify tests, we DON'T check for equality of 'next'.
+        return (
+            self.left == other.left
+            and self.right == other.right
+            and self.node == other.node
+        )
+
+    def __lt__(self, other):
+        return (self.node, self.left, self.right) < (
+            other.node,
+            other.left,
+            other.right,
+        )
+
+
+class SegmentList:
```

Comment:
Do we need this SegmentList structure now? It looks to me like we're only using linked lists of Segments now, so it would simplify things a lot if we didn't need this extra structure. Everything we can do to make things simpler and clearer here will knock multiple hours of effort and hundreds of lines off the C implementation!

Reply:
See comment below for why I have retained SegmentLists.